docs/proposals/202105256-cluster-class-and-managed-topologies.md

```diff
@@ -0,0 +1,480 @@
+    1. Sets the number of replicas on the control plane object from `spec.topology.controlPlane.replicas`.
+    1. Sets the k8s version on the control plane object from the `spec.topology.version`.
+    1. Add the following labels to the control plane object:
+       ```yaml
+        cluster.x-k8s.io/managed-topology: <cluster-name>
+       ```
+    1. Creates the control plane object.
+    1. Sets the `spec.controlPlaneRef` field for the Cluster object.
+    1. Saves the cluster object in the API server.
+5. Machine deployment object creation
+    1. For each `spec.topology.worker.deployments` item in the list
+    1. Create a name `<cluster-name>-<worker-set-name>` (if too long, hash it)
+    1. Initializes a new MachineDeployment object.
+    1. Sets the `clusterName` field on the MD object
+    1. Sets the `replicas` field on the MD object using `replicas` field for the set of worker nodes.
+    1. Sets the `version` field on the MD object from the `spec.topology.version`.
+    1. Sets the `spec.template.spec.bootstrap` on the MD object from the `ClusterClass.spec.worker.deployments[i].template.bootstrap.ref` field.
+    1. Sets the `spec.template.spec.infrastructureRef` on the MD object from the `ClusterClass.spec.worker.deployments[i].template.infrastructure.ref` field.
+    1. Generates the set of labels to be set on the MD object. The labels are additive to the class' labels list, and the value in the `spec.topology.worker.deployments[i].labels` takes precedence over any set by the ClusterClass. Also, include the managed topology label stated below:
+       ```yaml
+        cluster.x-k8s.io/managed-topology: <cluster-name>-<worker-set-name>
+       ```
+      Note: The managed topology label needs to be set on the individual Machine objects as well.
+    1. Creates the Machine Deployment object in the API server.
+
+![Creation of cluster with ClusterClass][./images/cluster-class/create.png]
+
+##### Update an existing Cluster using ClusterClass
+This section talks about updating a cluster which was created using a `ClusterClass` object.
+1. User updates the `cluster.spec.topology` field adhering to the update validation [criteria](#clusterclass-2).
+2. For the ControlPlane object in `spec.topology.controlPlane`, the cluster controller checks for the presence of the control plane object using the name `<cluster-name>`. If found,
+    1. Compares and updates the number of replicas, if necessary.
+    1. Compares and updates the k8s version, if necessary.
+    1. Updates the KCP object in the API server.
+3. The cluster controller reconciles the list of required machine deployments with the current list of managed machine deployments by:
+    1. Adding/Removing MachineDeployment if necessary.
+    1. Comparing and updating the number of replicas, if necessary.
+    1. Comparing and updating the k8s version for the MD, if necessary.
+    1. Updating the Machine Deployment object in the API server.
```

Comment:
Let's say that the cluster has a MachineDeployment that uses a specific machine image. Let's say that I want to update the MachineDeployment to use a different machine image, e.g., because I need to use a more recent OS version, where a CVE has been patched. However, I cannot upgrade the cluster's Kubernetes version, because I am on the latest available version.

How do I update the MachineDeployment's machine image?

If I do not use ClusterClass, then I create a new Infrastructure Template (e.g. AWSMachineTemplate if I'm using CAPA), and update the MachineDeployment's Infrastructure Reference. 

Assuming I understand the proposal correctly, if use ClusterClass, I cannot update the MachineDeployment's Infrastructure Reference, because ClusterClass defines the Reference, and the definition is immutable.

Reply:
Yes, that assumption is correct. This version of ClusterClass does not allow changes to the ClusterClass apart from adding new worker sets. As part of the future work, we have added a goal to allow cluster-specific transformations to change the values present in the ClusterClass. Once the baseline CC proposal is implemented, the plan is to move towards introducing these mutability options while defining the managed topology in the cluster.